User-based ObjectStore

VJalili opened the pull request:
_(1) Initially started at https://github.com/galaxyproject/galaxy/pull/4314; (2) all the commits of this PR were squashed into a single commit on Dec 11, 2019, the history of the changes are preserved via [this branch](https://github.com/VJalili/galaxy/tree/ubos_backup)._
## Introduction
This PR extends Galaxy's `ObjectStore` to enable users to _bring-their-own-resources_: users can _plug a media_ (e.g., Amazon S3 bucket) on which Galaxy will persist their datasets.

## Motivations
-	**unlimited storage:** users on Galaxy instances with limited storage resources (e.g., storage quota) can potentially have an unlimited storage by plugging their own (cloud-based) storage to Galaxy;
-	**data sharing:** having datasets generated by Galaxy stored on user’s cloud-based storage makes it easier sharing analysis results with collaborators;
-	**flexible persistence location:** members of different labs using a common Galaxy instance hosted at their institute can have their data stored on their lab’s network attached storage (NAS).

## Highlights
-	For users without a plugged storage media, Galaxy will continue to use an instance-wide configuration for their data storage needs;
-	A user's storage media (e.g., an S3 bucket) will be used for their data storage needs only, and will not be accessed for other user's storage needs; 
-	A _storage media_ can be a _local_ path, or an Amazon S3 bucket;
-	Users can plug multiple media (e.g., two different local path, and three Amazon S3 buckets), assign an `order` and `quota` attribute to each, and Galaxy will use them based on the given `order` and will fall from one to another if their quota limit is reached;
-	Leveraging the `order` attribute of _storage media_, users can use both instance-wide storage and their own media. For instance, they can direct Galaxy to use the instance-wide storage until their quota limit is reached (e.g., 250GB on Galaxy _Main_), then use their own media for the rest of their data storage needs. 
-	_Storage media_ are defined leveraging [Galaxy’s cloud authorization model]( https://doi.org/10.1093/bioinformatics/btz472), hence Galaxy does not ask for user’s credentials.
-	This PR implements all the necessary models, managers, functions, and APIs; and there will be a separate PR for UI;
-	The functionality is leveraging `Hierarchical` ObjectStore; hence, it is functional only if `Hierarchical` ObjectStore is configure. However, the hierarchy is applied instance-wide only, and does not affect user’s plugged media configuration;
-	Each _storage media_ has its separate staging path (mainly used for S3 backend), independent from the instance-wide ObjectStore and other storage media; and admin can define a default staging path.

## What's next?

We're aiming to keep this PR "minimally functional"; hence, features such as ability to mount a cloud-based storage and user interfaces will be implemented in subsequent PRs. 

## How to use

1. Configure objectstore to the hierarchical backend; e.g.,: 

```xml
<?xml version="1.0"?>
<object_store type="hierarchical">
    <backends>
        <object_store type="distributed" id="primary" order="0">
            <backends>
                <backend id="files1" type="disk" weight="1">
                    <files_dir path="database/files1"/>
                    <extra_dir type="temp" path="database/tmp1"/>
                    <extra_dir type="job_work" path="database/job_working_directory1"/>
                </backend>
                <backend id="files2" type="disk" weight="1">
                    <files_dir path="database/files2"/>
                    <extra_dir type="temp" path="database/tmp2"/>
                    <extra_dir type="job_work" path="database/job_working_directory2"/>
                </backend>
            </backends>
        </object_store>
        <object_store type="disk" id="secondary" order="1">
            <files_dir path="database/files3"/>
            <extra_dir type="temp" path="database/tmp3"/>
            <extra_dir type="job_work" path="database/job_working_directory3"/>
        </object_store>
    </backends>
</object_store>
```
2. Login and get your API key;
3. `POST` a payload as the following to the `/api/storage_media` (you may use Postman to send API requests):

```json
{
    "category": "local",
    "path": "A_PATH_ON_LOCAL_DISK",
    "order": "1",
    "quota": "1000.0",
    "usage": "0.0"
}
```

4. Then any dataset you create, will be stored in the `A_PATH_ON_LOCAL_DISK`; e.g.,:

```sh
.
└── d
    └── b
        └── 1
            └── dataset_db1b29ae-524a-46c1-af8d-e3e9e6861a4e.dat
```

jgoecks commented:
I started a branch with fixes here: https://github.com/jgoecks/galaxy/tree/UserBasedObjectStore2

Specifically, there are fixes for anonymous access. I can't seem to find your fork to initiate a pull request however—perhaps because your repo is restricted somehow and/or is so far behind the main repo?

VJalili commented:
Thanks for the updates @jgoecks . Please see if you can make a PR agains [this](https://github.com/VJalili/galaxy/tree/UserBasedObjectStore2) branch; if not, I can update this branch. Besides, I guess we could avoid your last commit.

jgoecks commented:
@VJalili I still cannot find your fork to make a PR against. I'll try to look into this more soon.

VJalili commented:
@jgoecks  I applied the changes you made on your branch on this branch.

qiagu commented:
It will be nice to have user-based storage. @VJalili Wonder whether you use sql tables to manage user and corresponding storages. I haven't looked deep into this project yet, but my first feeling is to build a table on top of current storage management system.

dannon commented:
@VJalili I opened a PR that I think will fix tests for this PR.  It's actually, I think, an issue we have always had and it was just never surfaced until this PR.

https://github.com/VJalili/galaxy/pull/6

VJalili commented:
@dannon Thank you! I guess that has fixed it as all tests passed locally.

VJalili commented:
@dannon I think the patch works fine for integration tests, but it breaks CI unit tests.

dannon commented:
@VJalili Ahh, sure enough.  I was laser focused on that one issue, let me dig deeper since there's more to the picture here.  

Yeah, the error here has popped up again:
`Parent instance <HistoryDatasetAssociation at 0x7fe8e45b0250> is not bound to a Session; lazy load operation of attribute 'history' cannot proceed`

I'll try to figure out how we're getting an hda handle that's no longer bound.

VJalili commented:
The orphan HDA handle is the issue causing the integration test's failure; I guess that is happening when Galaxy is writing metadata to a file.

martenson commented on test/integration/objectstore/test_plug_media.py:
```diff
@@ -309,7 +310,7 @@ def setUp(self):
         self.dataset_populator = DatasetPopulator(self.galaxy_interactor)
 
-    def test_media_selection_based_on_dataset_size(self):
+    def t__est_media_selection_based_on_dataset_size(self):
```
this does not seem right

VJalili replied:
Yes, I guess that was committed unintentionally, it will be fixed in subsequent commits.

jmchilton commented:
Thanks for refactoring the concept of ownership out of the dataset instance level (HDA/LDDA) and for the integration tests. These are serious improvements I believe. 

Can you add an integration test of copying data on storage media between users? I assume based on the reading if a user copies my data and then I delete the storage media - the data will disappear for the user but I want that verified and stated explicitly with a test case. Is that fair?

VJalili commented:
@jmchilton as per the challenges using this feature for shared data may introduce (e.g., authorization issues), last we decided to postpone the ability of using this feature for shared data. Do you think we should add some warnings for users who attempt to use this feature for share data?

jmchilton commented:
> Do you think we should add some warnings for users who attempt to use this feature for share data?

Yes, ideally. I'm not sure yet if that should be required for this PR but that is a good idea in general if we're going to impose that restriction.

VJalili commented:
@jmchilton I disabled sharing for user storage media (a history that contains a dataset stored on a user-owned storage, cannot be shared); please see [a272454](https://github.com/galaxyproject/galaxy/pull/4840/commits/a2724549c7fae7085318cf222708e8e0a94f8c5c). Any other thoughts?

jmchilton commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1402,6 +1429,7 @@ def _set_object_store_ids(self, job):
         # afterward. State below needs to happen the same way.
         for dataset_assoc in job.output_datasets + job.output_library_datasets:
             dataset = dataset_assoc.dataset
+            self.__assign_media(job, dataset.dataset)
```
I've spent months of my life trying to optimize this process of initializing the output datasets. Can we have some property on app that we can check to see if this method would ever doing anything - and skip it if there is no possibility of assigning media?

VJalili replied:
Sure! [this config property](https://github.com/galaxyproject/galaxy/pull/4840/files#diff-de4bb10f1af2a1d9d642c3904b86818cR1112) is added that [if disabled, this method will not do anything](https://github.com/galaxyproject/galaxy/pull/4840/files#diff-3c62e82e72ce988e633066c5a9b670bbR976). Would that be addressing your concerns?

jmchilton commented on lib/galaxy/config/__init__.py:
```diff
@@ -589,6 +589,17 @@ def __init__(self, **kwargs):
+        # Configuration options for new storage media uses can plug-in.
+        self.default_storage_media_jobs_directory = kwargs.get(
+            "default_storage_media_jobs_directory",
+            "database/job_working_directory_storage_media")
```
This is not how we use the database directory any more, I get that object_store_cache_path up there is also broken so I opened https://github.com/galaxyproject/galaxy/pull/8490 to try to fix it.

VJalili replied:
Good point, updated.

jmchilton commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1307,6 +1333,7 @@ def change_state(self, state, info=False, flush=True, job=None):
             return
         for dataset_assoc in job.output_datasets + job.output_library_datasets:
             dataset = dataset_assoc.dataset
+            model.StorageMedia.refresh_all_media_credentials(dataset.dataset.active_storage_media_associations, self.app.authnz_manager, self.sa_session)
```
This piece of code has also been heavily optimized - I'm really quite nervous about this extra association fetch per output. For huge collection outputs - this could result in a huge performance hit I would guess.

I'm not immediately sure how to handle this though. I already feel like this refresh around every operation is problematic - it seems like it should be push a layer down so these function calls don't have to worry about these details - but I'm not sure how to appropriately push the authnz_manager down into that level.

VJalili replied:
@jmchilton would https://github.com/VJalili/galaxy/commit/ca7e71609b5b08b7ee870e0452dd663b4e399372 be in the direction to address your performance hit concerns?  In short, it does not fetch db association or refresh credentials if the user-based objectstore feature is not enabled.

jmchilton commented on lib/galaxy/jobs/handler.py:
```diff
@@ -553,7 +553,14 @@ def __verify_job_ready(self, job, job_wrapper):
+        # exception(s).
+        if state == JOB_READY and self.app.config.enable_quotas and \
+                (job.user is not None and
+                 (job.user.active_storage_media is None or len(job.user.active_storage_media) == 0)):
```
I'd make this a property on job.user - ``has_active_storage_media`` or something to clean up this piece of code. This whole chunk though should probably be refactored into a ``check_user_over_quota`` method somewhere...

VJalili replied:
I implemented the property as you suggest. Reg. refactoring that chunk: I am leaving it as is for now for the sake of minimizing the changes with this PR, and we can modify as you suggested in subsequent PRs.

jmchilton commented on lib/galaxy/managers/hdas.py:
```diff
@@ -138,13 +138,20 @@ def purge(self, hda, flush=True):
         # decrease the user's space used
         if quota_amount_reduction:
-            user.adjust_total_disk_usage(-quota_amount_reduction)
+            if len(hda.dataset.active_storage_media_associations) == 0:
```
Again - needs to be a property - not a length check like this at this layer.

jmchilton commented on lib/galaxy/managers/hdas.py:
```diff
@@ -138,13 +138,20 @@ def purge(self, hda, flush=True):
+            if len(hda.dataset.active_storage_media_associations) == 0:
+                user.adjust_total_disk_usage(-quota_amount_reduction)
+            else:
+                for assoc in hda.dataset.active_storage_media_associations:
```
Why is it being subtracted from all of them?

VJalili replied:
It is reduced from all the storage media where the dataset was stored, because when purging a dataset, it is removed from all the media where it was stored.

jmchilton commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1918,6 +1953,7 @@ def setup_external_metadata(self, exec_dir=None, tmp_dir=None,
         if set_extension:
             for output_dataset_assoc in job.output_datasets:
                 if output_dataset_assoc.dataset.ext == 'auto':
+                    model.StorageMedia.refresh_all_media_credentials(output_dataset_assoc.dataset.dataset.active_storage_media_associations, self.app.authnz_manager, self.sa_session)
```
Setting the extension shouldn't require this - right? Something seems off here to me - maybe if I sat with it longer though it would become clear?

jmchilton commented on lib/galaxy/model/__init__.py:
```diff
@@ -1614,9 +1772,18 @@ def add_datasets(self, sa_session, datasets, parent_id=None, genome_build=None,
-            if quota and self.user:
-                disk_usage = sum([d.get_total_size() for d in datasets])
-                self.user.adjust_total_disk_usage(disk_usage)
+            if self.user:
```
This piece of code was highly optimized... I would be sad if this were merged without verifying there isn't a huge performance hit here.

VJalili replied:
How do you evaluate your performance gain? and how much reduction is acceptable?

VJalili replied:
I guess this point can be considered resolved as per discussion at https://github.com/galaxyproject/galaxy/issues/8790

jmchilton commented on lib/galaxy/objectstore/__init__.py:
```diff
@@ -577,12 +616,26 @@ def _repr_object_for_exception(self, obj):
+        if hasattr(obj, "active_storage_media_associations") and \
+                len(obj.active_storage_media_associations) > 0 and \
+                not ignore_media:
+            media = UserObjectStore(obj.active_storage_media_associations, self)
```
This may be fine but this piece of code repeated several times seems counter intuitive to me - shouldn't we be hitting the UserObjectStore if we need it and not overriding this stuff at this level in other object stores? How does the UserObjectStore interact with like DiskObjectStores then?

jmchilton replied:
I guess you addressed it above with 

> The functionality is leveraging Hierarchical ObjectStore; hence, it is functional only if Hierarchical ObjectStore is configure. However, the hierarchy is applied instance-wide only, and does not affect user’s plugged media configuration;

I don't know... I feel like the abstraction is subtly off here again though.

VJalili replied:
You mean we should not use Hierarchical objectstore? I guess @dannon finds this best,

blankenberg replied:
I think a point is to perhaps create a new class, could be a subclass of `HierarchicalObjectStore`, maybe `UserLocalHierarchicalObjectStore`, and have the 'special' behavior happen as a function of the new object?

VJalili replied:
Would the [UserObjectStore class](https://github.com/VJalili/galaxy/blob/6ecc54c56ab0a84b5876c07d11e3c2bd64463f17/lib/galaxy/objectstore/__init__.py#L965) be in the correct direction? The goal of the current implementation is to minimize the footprint, so that a new ObjectStore instance per user is **not** created at higher-up in the call hierarchy (e.g., at job execution level), rather a new instance creation per used is pushed to the ObjectStore level. This way the changes are largely limited to the one method of hierarchical ObjectStore only (i.e., [_call_method](https://github.com/VJalili/galaxy/blob/6ecc54c56ab0a84b5876c07d11e3c2bd64463f17/lib/galaxy/objectstore/__init__.py#L673)).

VJalili replied:
I follow up in https://github.com/galaxyproject/galaxy/pull/9554.

jmchilton commented on lib/galaxy/tools/actions/__init__.py:
```diff
@@ -413,7 +413,7 @@ def handle_output(name, output, hidden=None):
                 # Must flush before setting object store id currently.
                 trans.sa_session.flush()
-                object_store_populator.set_object_store_id(data)
+                object_store_populator.set_object_store_id(data=data, user=trans.user)
```
Seems like ObjectStorePopulator could be given the user when it is created and not need to changed down here. I think I would prefer that.

VJalili replied:
Done.

jmchilton commented on lib/galaxy/tools/parameters/basic.py:
```diff
@@ -1250,6 +1251,7 @@ def get_options(self, trans, other_values):
         if self.usecolnames:  # read first row - assume is a header with metadata useful for making good choices
             dataset = other_values.get(self.data_ref, None)
             try:
+                StorageMedia.refresh_all_media_credentials(dataset.active_storage_media_associations, self.app.authnz_manager, self.sa_session)
```
I mean ... if we do decide that we want to always call this "refresh" thing before we use the dataset methods, we can do a lot better in terms of abstractions I think. 

```
self.app.dataset_manager.ensure_dataset_ready(dataset)
```

would probably be a better abstraction. We wouldn't need to import StorageMedia, basic.py wouldn't need to know about dataset.active_storage_media_associations or authnz at all. ``app`` has its own ``self.sa_session`` so we wouldn't need to pass it. One argument versus three would be a more maintainable interface.

VJalili replied:
As per conversation, the cloud-based backends support are dropped for clarity. They will be made as a separate PR after this is merged.

jmchilton commented on lib/galaxy/webapps/galaxy/api/history_contents.py:
```diff
@@ -774,6 +775,7 @@ def __delete_dataset(self, trans, history_id, id, purge, **kwd):
         hda = self.hda_manager.get_owned(self.decode_id(id), trans.user, current_history=trans.history)
         self.hda_manager.error_if_uploading(hda)
 
+        model.StorageMedia.refresh_all_media_credentials(hda.dataset.active_storage_media_associations, trans.app.authnz_manager, trans.sa_session)
```
If we can't push it down below the manager level - then at very least this belongs in the manager level and not in the API level. I feel like ``purge`` and ``delete`` below should take care of these details if they are needed.

VJalili replied:
As per conversation at https://github.com/galaxyproject/galaxy/issues/8790 cloud-based backend support is removed from this.

jmchilton commented on lib/galaxy/webapps/galaxy/controllers/dataset.py:
```diff
@@ -975,7 +975,12 @@ def _purge(self, trans, dataset_id):
             # Decrease disk usage first
             if user:
-                user.adjust_total_disk_usage(-hda.quota_amount(user))
+                if len(hda.dataset.active_storage_media_associations) == 0:
```
See comment about API controller layer above. This shouldn't be happening in this controller I don't think.

VJalili replied:
Updated to use the property instead.

jmchilton commented:
This needs to be merged with dev so we can see the integration tests pass before it is merged. I'm not comfortable with merging it for 19.09 though - these abstractions still seem intrusive and off and I'm worried about performance hits in many critical paths this code touches.

I don't think I mentioned it in the inline comments but I feel like the quota pieces shouldn't be desperate - that if logic every place this touches quota should be moved into an existing or a new abstraction.

jmchilton commented:
My reading of the code is that it is at least still two big, largely independent chunks of work - one answering the question of how do we inject user credentials into the objectstore layer when needed and one answering a question of how do we switch between objectstores when some condition is active and manage quota in multiple places. I'd really still like to see this decomposed further into those two pieces. Again - that may not be possible or practical - but I think it would crystalize the review process and make it easier to provide concrete directions on working toward an implementation that... at least I... would feel comfortable merging?

VJalili commented:
How do you envision separating choosing/assigning a media from credentials?

VJalili commented:
Additionally, I guess  one of your main concerns is the performance penalty. Would it be possible to share your current performance evaluation results and methods, and set a threshold for performance penalty that we aim for?

jmchilton commented:
One can imagine many scenarios where one would want to dispatch between different object stores that has nothing to do with the cloud object store. 

- Two object stores are available with different quotas -  user opts to write new datasets to faster, less backed up disk to get more quota/faster performance for a short period of time.
- This same two object store scenario but that second one is used if the main quota/object store is full without user intervention (like this PR).
- This same two object store scenario but that second one is used if the main quota/object store and the user is on some access list.

I feel like the fact we'd like Galaxy to work in these different modalities means we need to provide some sort extension point, configuration point for how quota works and interacts with object stores.

I don't really feel anything in this PR is establishing abstractions that make this easier. I feel like you're modifying quota in different ways in different places and making it harder to do this. Alternatively - if we had a way to handle the above scenarios presumably this PR would glide right into that pretty naturally.

I'm sorry you feel like separating the quota pieces out would take months - I would suggest rebasing the whole PR into a single commit and then using ``git reset -p HEAD~`` to pull just the quota related pieces out. I'd recommend this video - https://johnkary.net/blog/git-add-p-the-most-powerful-git-feature-youre-not-using-yet/ - I use this technique daily and it is really great at breaking big PRs into smaller ones. I've broken out dozens of atomic PRs out of my remote data staging branch this way, and hundreds from my CWL branch.

VJalili committed:
Add User-based ObjectStore features.

VJalili commented:
> I would suggest rebasing the whole PR into a single commit

@jmchilton done. I will get to other points later.

jmchilton commented:
Thanks for the rebase! Pushing the milestone but hopefully we can make some substantial progress this next release cycle.

VJalili committed:
Drop AWS S3 support for User-Based ObjectStore:

VJalili commented:
> My reading of the code is that it is at least still two big, largely independent chunks of work - one answering the question of how do we inject user credentials into the objectstore layer when needed and ...

@jmchilton as you asked I removed any code related to credentials and AWS S3 support (the _media_ that requires authZ credentials), see https://github.com/VJalili/galaxy/commit/eda73341ef25f7e1872ce6b6d40dfd0cbfc09a0d.

> ... one answering a question of how do we switch between objectstores when some condition is active and manage quota in multiple places.

@jmchilton following your suggestion I removed all the quota-related logic; see https://github.com/VJalili/galaxy/commit/be3b31fe5111794320e756e83fa35c1717388fa6

I guess these changes are in alignment to the three points you have raised (i.e., remove quota, remove authnz, add an option to disable this feature for performance concerns). If these address the afore-mentioned points, I can next work on other comments and failing tests. Any suggestions?

VJalili committed:
Remove the option of auto user media selection:

preference/order on storage media;

VJalili commented:
ping @jmchilton

VJalili commented:
@jmchilton What is your path forward on this?

VJalili committed:
Change how working directory paths are resolved.

VJalili committed:
Use a property instead of a len check.

VJalili committed:
Pass user to the ObjectStore constructor rather than its methods.

VJalili committed:
Use the property instead of the len check.

VJalili committed:
Change the UBOS migration script number.

VJalili committed:
Replace more count length with accessing property.

VJalili committed:
Add the missing migration script.

VJalili committed:
Fix issues related to failing tests.

jdavcs commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -821,6 +821,42 @@ mapping:
+
+      default_storage_media_jobs_directory:
+        type: str
+        default: database/job_working_directory_storage_media
```
```suggestion
        default: job_working_directory_storage_media
        path_resolves_to: data_dir
```

VJalili replied:
Thanks.

jdavcs commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -821,6 +821,42 @@ mapping:
+
+      default_storage_media_cache_path:
+        type: str
+        default: database/storage_media_cache
```
```suggestion
        default: storage_media_cache
        path_resolves_to: data_dir
```

VJalili committed:
Reduce user count to 2 in order for tests to pass on CircleCI.

VJalili committed:
Remove database prefix from UBOS path.

blankenberg commented:
Here is what I did:

- Enable setting in galaxy.yml:
enable_user_based_object_store: true
- cp config/object_store_conf.xml.sample config/object_store_conf.xml
- Start galaxy.
- Post via curl: curl -H "Content-Type: application/json" -d '{"category": "local", "path": "/Volumes/shared/dan/galaxy-object-store-test/dan1", "order": "1", "quota": "1000.0", "usage": "0.0"}' 'http://localhost:8080/api/storage_media?key=apiKeyHere'

Some initial comments:

Lack of UI and clear documentation makes evaluation somewhat difficult, but I understand why not present.

Not sure why `enable_user_based_object_store: true` needs to exist, can’t this just come implicitly from the store being defined and enabled in the `object_store_conf.xml`?

It is not clear if I am able to (and if so, how) to define and enable some rulesets for this object store. It looks like I can just turn it on and off. If it is on, any body can try to create directories and files into any place that they want -- success/failure is based upon the permissions of the galaxy process user. There has to be a configurable way to limit/allow what can be done, and by whom.

First thing I might suggest is to consider how this all is being enabled/defined. My first inclination is to remove the new settings from galaxy.yml and extend `object_store_conf.xml`. Something perhaps like
```
        <object_store type="local_path">
             <filter type="allow_path"/>someThing</filter>
             <filter type="restrict_path"/>someThing</filter>
             <filter type="allow_users"/>someThing</filter>
             <config name="storage_media_jobs_directory" />database/job_working_directory_storage_media</>
             <config name="storage_media_cache_path" />database/storage_media_cache</>
             <config name="storage_media_cache_size" />100</>
...other config things here ...
        </object_store>
```
It is probably worthwhile to spend more time thinking about naming and structure, than what I just made up here.
 
I would then also consider what the config would look like for enabling/disabling/configuring the user-based Cloud storage. Then keep both of these use-cases in mind when reworking the code behavior. It could also be helpful to see the implementation of the cloud-based user store -- to demonstrate the abstractness and isolation of the system.

I’ve also added some comments inline with the code.


Please let me know if I can expand on anything. Can do a video call as well.

VJalili committed:
Merge remote-tracking branch 'upstream/dev' into UserBasedObjectStore2

# Conflicts:
#	lib/galaxy/config/__init__.py
#	lib/galaxy/objectstore/__init__.py
#	lib/galaxy/tools/actions/__init__.py

VJalili committed:
Increment migration script number.

VJalili committed:
Update to choose media at the _invoke method.

VJalili committed:
Remove `ignore_media`, and revert some cosmetic changes.

jdavcs commented on lib/galaxy/config/__init__.py:
```diff
@@ -576,6 +576,21 @@ def _process_config(self, kwargs):
+        self.object_store_cache_path = self.resolve_path(kwargs.get("object_store_cache_path", os.path.join(self.data_dir, "object_store_cache")))
+
+        # Configuration options for new storage media uses can plug-in.
+        self.enable_user_based_object_store = kwargs.get(
```
this is already done automatically (`self.enable_user_based_object_store` is created and by default initialized to `False`, as per schema

VJalili replied:
Thanks. I removed this and all the following ones.

jdavcs commented on lib/galaxy/config/__init__.py:
```diff
@@ -576,6 +576,21 @@ def _process_config(self, kwargs):
+        # Configuration options for new storage media uses can plug-in.
+        self.enable_user_based_object_store = kwargs.get(
+            "enable_user_based_object_store", False)
+        self.default_storage_media_jobs_directory = self.resolve_path(kwargs.get(
```
same as above

jdavcs replied:
also, you don't need to call `resolve_path` - that's already done as well

VJalili replied:
Thanks.

jdavcs commented on lib/galaxy/config/__init__.py:
```diff
@@ -576,6 +576,21 @@ def _process_config(self, kwargs):
+        self.default_storage_media_jobs_directory = self.resolve_path(kwargs.get(
+            "default_storage_media_jobs_directory",
+            os.path.join(self.data_dir, "job_working_directory_storage_media")))
+        self.default_storage_media_cache_path = self.resolve_path(kwargs.get(
```
same as above

jdavcs replied:
Also, `os.path.join` - same thing: all is done here: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L222 , then here: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L232

VJalili replied:
Thanks.

jdavcs commented on lib/galaxy/config/__init__.py:
```diff
@@ -576,6 +576,21 @@ def _process_config(self, kwargs):
+            "default_storage_media_cache_path",
+            os.path.join(self.data_dir, "storage_media_cache")))
+        self.default_storage_media_cache_size = kwargs.get(
+            "default_storage_media_cache_size",
```
same as above

VJalili replied:
Thanks.

VJalili committed:
Remove configurations that are auto-defined and set to their defaults.

VJalili committed:
Add storage media path to the test_config_values unit test.

nuwang commented on lib/galaxy/webapps/galaxy/api/storage_media.py:
```diff
@@ -0,0 +1,217 @@
+
+    @expose_api
+    def update(self, trans, encoded_media_id, payload, **kwargs):
+        msg_template = "Rejected user `" + str(trans.user.id) + "`'s request to updade storage media config because of {}."
```
```suggestion
        msg_template = "Rejected user `" + str(trans.user.id) + "`'s request to update storage media config because of {}."
```

VJalili replied:
Thanks.

nuwang commented on lib/galaxy/webapps/galaxy/api/storage_media.py:
```diff
@@ -0,0 +1,217 @@
+        try:
+            usage = float(payload.get("usage", "0.0"))
+        except ValueError:
+            return "Expect a float number for the `usage` attribute, but received `{}`.".format(payload.get("usage"))
```
```suggestion
            return "Expected a floating-point number for the `usage` attribute, but received `{}`.".format(payload.get("usage"))
```

VJalili replied:
Thank you.

VJalili committed:
Update lib/galaxy/webapps/galaxy/api/storage_media.py

Co-authored-by: Nuwan Goonasekera <2070605+nuwang@users.noreply.github.com>

jmchilton commented on lib/galaxy/webapps/galaxy/controllers/history.py:
```diff
@@ -1057,7 +1063,12 @@ def purge_deleted_datasets(self, trans):
                 if trans.user:
-                    trans.user.adjust_total_disk_usage(-hda.quota_amount(trans.user))
+                    if not hda.dataset.has_active_storage_media():
+                        trans.user.adjust_total_disk_usage(-hda.quota_amount(trans.user))
```
I think we can get around duplicating this code in the controllers with https://github.com/galaxyproject/galaxy/pull/10208.

jmchilton replied:
That other PR has been merged so I think this should be rebased now along with the fix for https://github.com/galaxyproject/galaxy/pull/4840/files#r486551671.

jmchilton commented on lib/galaxy/model/__init__.py:
```diff
@@ -2602,8 +2713,8 @@ def get_size(self, nice_size=False):
-        """Sets and gets the size of the data on disk"""
-        return self.dataset.set_size(**kwds)
+        """Sets the size of the data on disk"""
+        self.dataset.set_size(**kwds)
```
Is this a broken rebase or is it important to drop the return here?

jmchilton commented on lib/galaxy/tools/actions/__init__.py:
```diff
@@ -359,7 +359,7 @@ def execute(self, tool, trans, incoming=None, return_job=False, set_output_hid=T
         parent_to_child_pairs = []
         child_dataset_names = set()
-        object_store_populator = ObjectStorePopulator(app)
+        object_store_populator = ObjectStorePopulator(app, user=trans.user)
```
Can you open a new PR with these changes - https://github.com/galaxyproject/galaxy/compare/dev...jmchilton:user_objectstore_populator?expand=1. This continues a theme along with https://github.com/galaxyproject/galaxy/pull/10208 and https://github.com/galaxyproject/galaxy/pull/10212 of trying to establish Galaxy abstractions that restrict the code needed to implement this functionality just to object store, quota, and model code.

VJalili replied:
Sure: https://github.com/galaxyproject/galaxy/pull/10231

jmchilton commented on lib/galaxy/webapps/galaxy/controllers/history.py:
```diff
@@ -969,7 +969,13 @@ def _populate_restricted(self, trans, user, histories, send_to_users, action, se
                     # Only deal with datasets that have not been purged
                     for hda in history.activatable_datasets:
-                        if trans.app.security_agent.can_access_dataset(send_to_user.all_roles(), hda.dataset):
+                        if len(hda.dataset.storage_media_associations) > 0:
```
This is not sufficient to prevent sharing at all I don't think. There are other paths to share datasets that don't hit this controller, library datasets should be prohibited from being such datasets, etc...

I think https://github.com/galaxyproject/galaxy/pull/10840 is what we want. It is much more general - it allows any objectstore to be marked as private - and it is much more comprehensive in how it prevents sharing. It has test cases, it prevents such datasets from even showing up where say importing history datasets into libraries, etc...

I think this portion of the PR should be dropped when that other PR is merged and instead just ensure that your user based objectstores are marked as private - I think better APIs and UIs will pretty cleanly fallout from that.

jmchilton commented on lib/galaxy/jobs/handler.py:
```diff
@@ -558,7 +558,14 @@ def __verify_job_ready(self, job, job_wrapper):
+        # exception(s).
+        if state == JOB_READY and self.app.config.enable_quotas and \
+                (job.user is not None and
+                 (job.user.active_storage_media is None or not job.user.has_active_storage_media())):
```
I think this should be redone on top of https://github.com/galaxyproject/galaxy/pull/10221 - which I think abstracts out the quota checking into nice optimizable functions. I think rather than checking if has_active_storage_media we should build on the abstractions in that PR to just ask if the configured objectstore we're talking to has quota left and then we can disable quota on objectstores that use storage media. The ability to disable quota on an objectstore is included in that PR.
